soc/bouffalolab/bl60x/soc.c

```diff
@@ -0,0 +1,890 @@
+	tmpVal  = sys_read32(GLB_BASE + GLB_CLK_CFG0_OFFSET);
+	tmpVal  = (tmpVal & GLB_REG_PLL_SEL_UMSK)
+		| ((uint32_t)(clock_frequency_source - 1) << GLB_REG_PLL_SEL_POS);
+	sys_write32(tmpVal, GLB_BASE + GLB_CLK_CFG0_OFFSET);
+
+	/* set root clock settings */
+	switch (clock_frequency_source) {
+	case 32:
+		/* we are not supposed to be here */
+		break;
+	case 0:
+		system_set_root_clock(1);
+		sys_write32(system_uart_bflb_get_crystal_frequency(), CORECLOCKREGISTER);
+		break;
+	case 1:
+		system_set_root_clock(crystal == 32 ? 2 : 3);
+		sys_write32(MHZ(48), CORECLOCKREGISTER);
+		break;
+	case 2:
+		system_set_root_clock_dividers(0, 1);
+		system_set_root_clock(crystal == 32 ? 2 : 3);
+		sys_write32(MHZ(120), CORECLOCKREGISTER);
+		break;
+	case 3:
+		/* TODO: enable rom access 2T*/
+		system_set_root_clock_dividers(0, 1);
+		system_set_root_clock(crystal == 32 ? 2 : 3);
+		sys_write32(MHZ(160), CORECLOCKREGISTER);
+		break;
+	case 4:
+		/* TODO: enable rom access 2T*/
+		system_set_root_clock_dividers(0, 1);
+		system_set_root_clock(crystal == 32 ? 2 : 3);
+		sys_write32(MHZ(192), CORECLOCKREGISTER);
+		break;
+	default:
+		break;
+	}
+
+	/* settle */
+	system_clock_delay_32M_ms(10);
+
+	/* we have PLL now */
+	system_set_PKA_clock(1);
+}
+
+static void uart_set_clock_enable(uint32_t enable)
+{
+	uint32_t tmpVal = 0;
+
+	if (enable) {
+		enable = 1;
+	}
+
+	tmpVal  = sys_read32(GLB_BASE + GLB_CLK_CFG2_OFFSET);
+	tmpVal  = (tmpVal & GLB_UART_CLK_EN_UMSK)
+		| (enable << GLB_UART_CLK_EN_POS);
+	sys_write32(tmpVal, GLB_BASE + GLB_CLK_CFG2_OFFSET);
+}
+
+/* Clock:
+ * FCLK: 0
+ * 160 Mhz PLL: 1
+ * When using PLL root clock, we can use either setting, when using the 32Mhz
+ * Oscillator with a uninitialized PLL, only FCLK will be available.
+ */
+static void uart_set_clock(uint32_t enable, uint32_t clock, uint32_t divider)
+{
+	uint32_t tmpVal = 0;
+
+	if (divider > 0x7) {
+		divider = 0x7;
+	}
+
+	if (clock) {
+		clock = 1;
+	}
+	/* disable uart clock */
+	uart_set_clock_enable(0);
+
+	tmpVal  = sys_read32(GLB_BASE + GLB_CLK_CFG2_OFFSET);
+	tmpVal  = (tmpVal & GLB_UART_CLK_DIV_UMSK)
+		| (divider << GLB_UART_CLK_DIV_POS);
+	sys_write32(tmpVal, GLB_BASE + GLB_CLK_CFG2_OFFSET);
+
+	tmpVal  = sys_read32(HBN_BASE + HBN_GLB_OFFSET);
+	tmpVal  = (tmpVal & HBN_UART_CLK_SEL_UMSK)
+		| (clock << HBN_UART_CLK_SEL_POS);
+	sys_write32(tmpVal, HBN_BASE + HBN_GLB_OFFSET);
+
+	uart_set_clock_enable(enable);
+}
+
+
+/* TODO: should take crystal type, defaults to 40Mhz crystal for BL602 as seems
+ * the most common
+ */
+#define USE_CRYSTALL_32MHZ 0
+static void system_clock_init(void)
+{
+#if USE_CRYSTALL_32MHZ
+	system_init_root_clock(0, 32);
+	system_set_root_clock_dividers(0, 0);
+	system_clock_trim_32M();
+#else
+	system_init_root_clock(5, 4);
+	system_set_root_clock_dividers(0, 2);
+#endif
+	system_set_machine_timer_clock(1, 0, mtimer_get_clk_src_div());
+}
+
+static void peripheral_clock_init(void)
+{
+	uint32_t regval = sys_read32(0x40000024);
```

Comment:
Magic address? We should probably get it from the devicetree. See if e.g. the syscon driver would fit here.

Reply:
This is an undocumented Global Control Register (GLB) register. We are planning to fix this after #72102 because we want to use directly the newer API. We will map later the peripherals index to this regions and fix it properly.

Reply:
Fixed GLB_BASE + GLB_CGEN_CFG1_OFFSET
BTW, It is undocumented.